src/main/scala/mesosphere/mesos/Constraints.scala

```diff
@@ -125,4 +126,85 @@ object Constraints {
+    */
+  //scalastyle:off return
+  def selectTasksToKill(app: AppDefinition, runningTasks: Set[MarathonTask], toKillCount: Int): Set[MarathonTask] = {
+    require(toKillCount <= runningTasks.size, "Can not kill more instances than running")
+
+    //short circuit, if all tasks shall be killed
+    if (runningTasks.size == toKillCount) return runningTasks
+
+    //currently, only the GROUP_BY operator is able to select tasks to kill
+    val distributions = app.constraints.filter(_.getOperator == Operator.GROUP_BY).map { constraint =>
+      def groupFn(task: MarathonTask): Option[String] = constraint.getField match {
+        case "hostname"    => Some(task.getHost)
+        case field: String => task.getAttributesList.asScala.find(_.getName == field).map(_.getText.getValue)
+      }
+      GroupByDistribution(constraint, runningTasks.groupBy(groupFn).values.toSeq)
+    }
+
+    //short circuit, if there are no constraints to align with
+    if (distributions.isEmpty) return Set.empty
+
+    var toKillTasks = Set.empty[MarathonTask]
+    var flag = true
+    while (flag && toKillTasks.size != toKillCount) {
+      val tried =
+        distributions.toSeq.sortBy(_.distributionDifference(toKillTasks)).reverseIterator //highest diff first
+          .flatMap(_.selectTaskToKill(toKillTasks)) ++ //select all tasks to kill
+          runningTasks.iterator.filterNot(toKillTasks.contains) // fallback: add all tasks not visited
+
+      tried.find(tryTask => distributions.forall(_.isMoreEvenWithout(toKillTasks + tryTask))) match {
+        case Some(task) => toKillTasks += task
+        case None       => flag = false
+      }
+    }
+
+    //log the selected tasks and why they were selected
+    if (log.isInfoEnabled) {
+      val taskDesc = toKillTasks.map { task =>
+        val attrs = task.getAttributesList.asScala.map(a => s"${a.getName}=${a.getText.getValue}").mkString(", ")
+        s"Task:${task.getId} host:${task.getHost} attrs:$attrs"
+      }.mkString("Selected Tasks to kill:\n", "\n", "\n")
+      val distDesc = distributions.map { d =>
+        val (before, after) = (d.distributionDifference(), d.distributionDifference(toKillTasks))
+        s"${d.constraint.getField} changed from: $before to $after"
+      }.mkString("Selected Constraint diff changed:\n", "\n", "\n")
+      log.info(s"$taskDesc$distDesc")
+    }
+
+    toKillTasks
+  }
+
+  /**
+    * Helper class for easier distribution computation.
+    */
+  private case class GroupByDistribution(constraint: Constraint, distribution: Seq[Set[MarathonTask]]) {
+    def isMoreEvenWithout(selected: Set[MarathonTask]): Boolean = {
+      val diffAfterKill = distributionDifference(selected)
+      //diff after kill is 0=perfect, 1=tolerated or minimizes the difference
+      diffAfterKill <= 1 || distributionDifference() > diffAfterKill
+    }
+
+    def selectTaskToKill(without: Set[MarathonTask], butNot: Set[MarathonTask] = Set.empty): Iterator[MarathonTask] = {
```

Comment:
To make it easy to read, I'd add a description of what the method does:

Removes from distribution the tasks in the "without"  parameter and returns:

- An iterator over an empty sequence if the distribution is even.
- An iterator over all the tasks belonging to the value with the largest amount of members otherwise.